File: lab3/s1/web/src/main/java/ru/uvuv643/web/HumanWebServiceImpl.java
package ru.uvuv643.web;

import jakarta.validation.Valid;
import jakarta.validation.constraints.PositiveOrZero;
import jakarta.ws.rs.*;
import jakarta.ws.rs.core.MediaType;
import jakarta.ws.rs.core.Response;
import ru.uvuv643.ejb.remote.RemoteServiceEjb;
import ru.uvuv643.ejb.remote.dto.enums.SortOrderDto;
import ru.uvuv643.ejb.remote.dto.enums.StatisticFieldDto;
import ru.uvuv643.ejb.remote.dto.enums.StatisticOperationDto;
import ru.uvuv643.ejb.remote.dto.human.request.CreateHumanBeingRequest;
import ru.uvuv643.ejb.remote.dto.human.request.ModifyHumanBeingRequest;
import ru.uvuv643.web.config.ConsulConfig;

import javax.naming.InitialContext;
import javax.naming.NamingException;
import java.util.List;

@Consumes(MediaType.APPLICATION_XML)
@Produces(MediaType.APPLICATION_XML)
@Path("/v1/")
public class HumanWebServiceImpl implements HumanWebService {

    private final RemoteServiceEjb remoteServiceEjb;

    public HumanWebServiceImpl() throws NamingException {
        this.remoteServiceEjb = this.getFromEJBPool();
    }

    @Override
    @GET
    @Path("/human-being")
    public Response getAllHumanBeing(
            @QueryParam("idGte") Integer idGte,
            @QueryParam("idLte") Integer idLte,
            @QueryParam("nameIn") List<String> nameIn,
            @QueryParam("coordinateXGte") Integer coordinateXGte,
            @QueryParam("coordinateXLte") Integer coordinateXLte,
            @QueryParam("coordinateYGte") Long coordinateYGte,
            @QueryParam("coordinateYLte") Long coordinateYLte,
            @QueryParam("creationDateGte") String creationDateGte,
            @QueryParam("creationDateLte") String creationDateLte,
            @QueryParam("realHero") Boolean realHero,
            @QueryParam("hasToothpick") Boolean hasToothpick,
            @QueryParam("impactSpeedGte") Double impactSpeedGte,
            @QueryParam("impactSpeedLte") Double impactSpeedLte,
            @QueryParam("minutesOfWaitingGte") Double minutesOfWaitingGte,
            @QueryParam("minutesOfWaitingLte") Double minutesOfWaitingLte,
            @QueryParam("coolCar") Boolean coolCar,
            @QueryParam("moodIn") List<String> moodIn,
            @QueryParam("weaponTypeIn") List<String> weaponTypeIn,
            @QueryParam("page") Integer page,
            @QueryParam("size") Integer size,
            @QueryParam("sortFields") List<String> sortFields,
            @QueryParam("sortDirections") List<SortOrderDto> sortDirections) {

        return Response.ok(remoteServiceEjb.getAllHumanBeing(idGte, idLte, nameIn, coordinateXGte, coordinateXLte, coordinateYGte,
                coordinateYLte, creationDateGte, creationDateLte, realHero, hasToothpick, impactSpeedGte, impactSpeedLte,
                minutesOfWaitingGte, minutesOfWaitingLte, coolCar, moodIn, weaponTypeIn, page, size, sortFields, sortDirections)).header("X-Server", ConsulConfig.service).build();
    }

    @Override
    @Consumes(MediaType.APPLICATION_XML)
    @Produces(MediaType.APPLICATION_XML)
    @POST
    @Path("/human-being")
    public Response createHumanBeing(@Valid CreateHumanBeingRequest request) {
        remoteServiceEjb.createHumanBeing(request);
        return Response.ok().header("X-Server", ConsulConfig.service).build();
    }

    @Override
    @GET
    @Path("/human-being/{id}")
    @Produces(MediaType.APPLICATION_XML)
    public Response getHumanBeingById(@PathParam("id") int id) {
        return Response.ok(remoteServiceEjb.getHumanBeingById(id)).header("X-Server", ConsulConfig.service).build();
    }

    @Override
    @PUT
    @Path("/human-being/{id}")
    @Consumes(MediaType.APPLICATION_XML)
    @Produces(MediaType.APPLICATION_XML)
    public Response modifyHumanBeing(@PathParam("id") String id, @Valid ModifyHumanBeingRequest request) {
        remoteServiceEjb.modifyHumanBeing(id, request);
        return Response.ok().header("X-Server", ConsulConfig.service).build();
    }

    @Override
    @GET
    @Path("/human-being/stats")
    @Produces(MediaType.APPLICATION_XML)
    public Response getHumanStats( @QueryParam("field") StatisticFieldDto field,
                                   @QueryParam("operation") StatisticOperationDto operation) {
        return Response.ok(remoteServiceEjb.getHumanStats(field, operation)).header("X-Server", ConsulConfig.service).build();
    }

    @DELETE
    @Path("/human-being")
    @Produces(MediaType.APPLICATION_XML)
    @Override
    public Response deleteByParams(
            @QueryParam("id") Integer id,
            @QueryParam("carCool") Boolean carCool,
                @QueryParam("impactSpeed") Double impactSpeed,
            @QueryParam("limit") @PositiveOrZero @Valid Integer limit
    ) {
        remoteServiceEjb.deleteByParams(id, carCool, impactSpeed, limit);
        return Response.ok().header("X-Server", ConsulConfig.service).build();
    }

    private RemoteServiceEjb getFromEJBPool() throws NamingException {
        return (RemoteServiceEjb) new InitialContext().lookup("java:global/remote-1.0-SNAPSHOT/ServiceEjbImpl!ru.uvuv643.ejb.remote.RemoteServiceEjb");
    }

}

File: lab1/s1/src/main/kotlin/ru/uvuv643/soa/server/HumanWebServiceImpl.kt
package ru.uvuv643.soa.server

import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.ResponseBody
import ru.uvuv643.soa.api.v1.HumanWebService
import ru.uvuv643.soa.api.v1.dto.enums.*
import ru.uvuv643.soa.api.v1.dto.human.HumanBeingDto
import ru.uvuv643.soa.api.v1.dto.human.ListHumanBeingDto
import ru.uvuv643.soa.api.v1.dto.human.request.CreateHumanBeingRequest
import ru.uvuv643.soa.api.v1.dto.human.request.DeleteHumanBeingRequest
import ru.uvuv643.soa.api.v1.dto.human.request.ModifyHumanBeingRequest
import ru.uvuv643.soa.api.v1.dto.human.response.DeleteResponseDto
import ru.uvuv643.soa.api.v1.dto.human.response.StatisticResponseDto
import java.util.*

@Controller
class HumanWebServiceImpl : HumanWebService {
    override fun getAllHumanBeing(
        idGte: Optional<Int>,
        idLte: Optional<Int>,
        nameIn: Optional<List<String>>,
        coordinateXGte: Optional<Int>,
        coordinateXLte: Optional<Int>,
        coordinateYGte: Optional<Double>,
        coordinateYLte: Optional<Double>,
        creationDateGte: Optional<Date>,
        creationDateLte: Optional<Date>,
        realHero: Optional<Boolean>,
        hasToothpick: Optional<Boolean>,
        impactSpeedGte: Optional<Float>,
        impactSpeedLte: Optional<Float>,
        minutesOfWaitingGte: Optional<Long>,
        minutesOfWaitingLte: Optional<Long>,
        coolCar: Optional<Boolean>,
        moodIn: Optional<List<MoodDto>>,
        weaponTypeIn: Optional<List<WeaponTypeDto>>,
        page: Optional<Int>,
        size: Optional<Int>,
        sortDirections: List<String>,
        sortFields: List<String>
    ): ListHumanBeingDto {
        TODO("Not yet implemented")
    }

    override fun createHumanBeing(request: CreateHumanBeingRequest): HumanBeingDto {
        TODO("Not yet implemented")
    }

    override fun getHumanBeingById(id: Int): HumanBeingDto {
        TODO("Not yet implemented")
    }

    override fun modifyHumanBeing(id: String, request: ModifyHumanBeingRequest): HumanBeingDto {
        TODO("Not yet implemented")
    }


    override fun getHumanStats(field: Optional<StatisticFieldDto>, operation: Optional<StatisticOperationDto>): StatisticResponseDto {
        TODO("Not yet implemented")
    }

    override fun deleteByParams(carCool: Boolean?, impactSpeed: Float?, limit: Int?): ResponseBody {
        TODO("Not yet implemented")
    }


}
